### ISSUE TITLE
Different behavior when running with -directory vs. on a specific file

### ISSUE THREAD

**Issue by sqs** [comby-tools/comby#73]
I am seeing Comby omit matches/replacements on some files in a directory specified with `-directory`. It finds them if the file is specified explicitly. See the repro script below.

```
$ comby -version
0.7.0
$ uname -a
Linux xq 4.18.0-25-generic #26-Ubuntu SMP Mon Jun 24 09:32:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

## Repro script

```bash
#!/bin/bash

set -eu

tmpdir=$(mktemp -d)
cd $tmpdir
curl -SL https://github.com/sourcegraph/sourcegraph/archive/ff38872f93cc0b5b9a21f56b919621e02e7dc800.tar.gz | tar xzf -
cd sourcegraph-ff38872f93cc0b5b9a21f56b919621e02e7dc800

echo
echo '### Running against the whole directory produces some matches:'
echo

comby -extensions go -directory cmd/frontend/graphqlbackend/ ':[[name]]At() string { return :[v].:[[name]]At.Format(time.RFC3339) }' ':[name]At() DateTime { return DateTime{:[v].:[name]At} }'

echo
echo '### But running it against a specific file in the directory produces matches in that file that were not produced (as expected) by the command above'
echo

comby ':[[name]]At() string { return :[v].:[[name]]At.Format(time.RFC3339) }' ':[name]At() DateTime { return DateTime{:[v].:[name]At} }' cmd/frontend/graphqlbackend/user.go
```

## Output of repro script

```
xq:~ $ bash ~/tmp/comby-repro.bash 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   165    0   165    0     0    435      0 --:--:-- --:--:-- --:--:--   435
100 10.4M  100 10.4M    0     0  8958k      0  0:00:01  0:00:01 --:--:-- 21.2M
/tmp/tmp.vCLqQUXHYa/sourcegraph-ff38872f93cc0b5b9a21f56b919621e02e7dc800

### Running against the whole directory produces some matches:

------ cmd/frontend/graphqlbackend/access_token.go
++++++ cmd/frontend/graphqlbackend/access_token.go
@|-60,9 +60,7 ============================================================
 |	return UserByIDInt32(ctx, r.accessToken.CreatorUserID)
 |}
 |
-|func (r *accessTokenResolver) CreatedAt() string {
-|	return r.accessToken.CreatedAt.Format(time.RFC3339)
-|}
+|func (r *accessTokenResolver) CreatedAt() DateTime { return DateTime{r.accessToken.CreatedAt} }
 |
 |func (r *accessTokenResolver) LastUsedAt() *string {
 |	if r.accessToken.LastUsedAt == nil {
------ cmd/frontend/graphqlbackend/org_members.go
++++++ cmd/frontend/graphqlbackend/org_members.go
@|-45,10 +45,6 ============================================================
 |	return UserByIDInt32(ctx, r.membership.UserID)
 |}
 |
-|func (r *organizationMembershipResolver) CreatedAt() string {
-|	return r.membership.CreatedAt.Format(time.RFC3339)
-|}
+|func (r *organizationMembershipResolver) CreatedAt() DateTime { return DateTime{r.membership.CreatedAt} }
-|
-|func (r *organizationMembershipResolver) UpdatedAt() string {
-|	return r.membership.UpdatedAt.Format(time.RFC3339)
-|}
+|
+|func (r *organizationMembershipResolver) UpdatedAt() DateTime { return DateTime{r.membership.UpdatedAt} }
------ cmd/frontend/graphqlbackend/org.go
++++++ cmd/frontend/graphqlbackend/org.go
@|-80,7 +80,7 ============================================================
 |
 |func (o *OrgResolver) SettingsURL() *string { return strptr(o.URL() + "/settings") }
 |
-|func (o *OrgResolver) CreatedAt() string { return o.org.CreatedAt.Format(time.RFC3339) }
+|func (o *OrgResolver) CreatedAt() DateTime { return DateTime{o.org.CreatedAt} }
 |
 |func (o *OrgResolver) Members(ctx context.Context) (*staticUserConnectionResolver, error) {
 |	// 🚨 SECURITY: Only org members can list the org members.
------ cmd/frontend/graphqlbackend/external_service.go
++++++ cmd/frontend/graphqlbackend/external_service.go
@|-68,13 +68,9 ============================================================
 |	return r.externalService.Config
 |}
 |
-|func (r *externalServiceResolver) CreatedAt() string {
-|	return r.externalService.CreatedAt.Format(time.RFC3339)
-|}
+|func (r *externalServiceResolver) CreatedAt() DateTime { return DateTime{r.externalService.CreatedAt} }
-|
-|func (r *externalServiceResolver) UpdatedAt() string {
-|	return r.externalService.UpdatedAt.Format(time.RFC3339)
-|}
+|
+|func (r *externalServiceResolver) UpdatedAt() DateTime { return DateTime{r.externalService.UpdatedAt} }
 |
 |func (r *externalServiceResolver) Warning() *string {
 |	if r.warning == "" {
------ cmd/frontend/graphqlbackend/survey_response.go
++++++ cmd/frontend/graphqlbackend/survey_response.go
@|-46,9 +46,7 ============================================================
 |	return s.surveyResponse.Better
 |}
 |
-|func (s *surveyResponseResolver) CreatedAt() string {
-|	return s.surveyResponse.CreatedAt.Format(time.RFC3339)
-|}
+|func (s *surveyResponseResolver) CreatedAt() DateTime { return DateTime{s.surveyResponse.CreatedAt} }
 |
 |// SurveySubmissionInput contains a satisfaction (NPS) survey response.
 |type SurveySubmissionInput struct {
------ cmd/frontend/graphqlbackend/org_invitation.go
++++++ cmd/frontend/graphqlbackend/org_invitation.go
@|-52,7 +52,7 ============================================================
 |func (r *organizationInvitationResolver) Recipient(ctx context.Context) (*UserResolver, error) {
 |	return UserByIDInt32(ctx, r.v.RecipientUserID)
 |}
-|func (r *organizationInvitationResolver) CreatedAt() string { return r.v.CreatedAt.Format(time.RFC3339) }
+|func (r *organizationInvitationResolver) CreatedAt() DateTime { return DateTime{r.v.CreatedAt} }
 |func (r *organizationInvitationResolver) NotifiedAt() *string {
 |	if r.v.NotifiedAt == nil {
 |		return nil

### But running it against a specific file in the directory produces matches in that file that were not produced (as expected) by the command above

------ /tmp/tmp.vCLqQUXHYa/sourcegraph-ff38872f93cc0b5b9a21f56b919621e02e7dc800/cmd/frontend/graphqlbackend/user.go
++++++ /tmp/tmp.vCLqQUXHYa/sourcegraph-ff38872f93cc0b5b9a21f56b919621e02e7dc800/cmd/frontend/graphqlbackend/user.go
@|-124,9 +124,7 ============================================================
 |
 |func (r *UserResolver) SettingsURL() *string { return strptr(r.URL() + "/settings") }
 |
-|func (r *UserResolver) CreatedAt() string {
-|	return r.user.CreatedAt.Format(time.RFC3339)
-|}
+|func (r *UserResolver) CreatedAt() DateTime { return DateTime{r.user.CreatedAt} }
 |
 |func (r *UserResolver) UpdatedAt() *string {
 |	t := r.user.UpdatedAt.Format(time.RFC3339) // ISO
```

**Comment by rvantonder** [comment]
The problem is with the `-extensions` flag: this expects the dot separator (`.go`). In the current behavior, because the dot is absent, the matcher silently falls back to the generic one, which doesn't find that match in `user.go`. It also explains why the explicit file option works: in this case, the extension is inferred. 

So, simply: 

```
comby -extensions .go -directory cmd/frontend/graphqlbackend/ ':[[name]]At() string { return :[v].:[[name]]At.Format(time.RFC3339) }' ':[name]At() DateTime { return DateTime{:[v].:[name]At} }' -stats
```

which shows 12 matches, including the missing one:
```
{
  "number_of_files": 135,
  "lines_of_code": 25914,
  "number_of_matches": 12,
  "total_time": 991.2979602813721
}
```
instead of 8 in the original command (if you tag on `-stats`).

It may make sense to be less strict about requiring the "." separator--I notice that codemod omits it. Happy to take suggestions.

**Comment by sqs** [comment]
Ah, I see. Thanks for the quick reply! My immediate issue is now resolved, so this issue is purely about UX now. 

1. Since the extensions flag docs mention it supports suffixes (the main.go example), I assumed that just go (no dot) would work. And indeed I was used to how codemod works. 
2. It is unexpected to me that file extension inference would occur for explicitly listed files but not for files in a directory I specify.
3. If it falls back to using the generic matcher, that seems worthy of at least a default stderr log message. That message would likely have helped me avoid this problem here.

I actually prefer how fastmod CLI args work (vs codemod, and admittedly with less experience comby). Fastmod lets you put all the files you want at the end like grep. The first 2 positional args are the pattern and replacement. This is nice because then I can do whatever $(...) expression using find, etc., I want to specify my list of files, and I don’t need to learn (or be limited by) the flags of the particular tool I am using. (Perhaps this is how comby works...I’m not at my computer to confirm, but iirc it didn’t accept multiple filenames listed after the patterns.)

So that (fastmod/grep-style) would be my drive-by, “as-a-user” suggestion, along with continuing to do file extension inference on explicitly specified files, warning if the generic syntax is used, and removing all -directory, -extensions, -r, etc., flags to avoid confusion.

**Comment by rvantonder** [comment]
> 1. Since the extensions flag docs mention it supports suffixes (the main.go example), I assumed that just go (no dot) would work. And indeed I was used to how codemod works.

Yeah, it does support suffixes (`go` is used as a suffix in the example, so will match `.go` files, but also, e.g., `bar.foogo`). Currently, neither case will cause the go matcher be inferred. While it might make sense to infer a `.go` file, it'd be better to fall back to a generic matcher if the intention was to match a `.foogo` file.

> 2. It is unexpected to me that file extension inference would occur for explicitly listed files but not for files in a directory I specify.

Fair. The heuristic being what it is, it's clear from the command line when a dot is omitted or not (and thus whether this triggers inference). If matchers are inferred during a directory traversal, there's some hidden decisions being made about how files are parsed. As a solution, (and to point (3)): I mostly see this as a problem with not being explicit about when matchers are inferred, and it's a bit conflated with the mechanism for filtering files. While there's a `-matcher` option to make the matcher explicit and override inference, the default behavior (with `-matcher` omitted) can clearly be improved. A loud message seems like the right idea.

> The first 2 positional args are the pattern and replacement. This is nice because then I can do whatever $(...) expression using find, etc., I want to specify my list of files, and I don’t need to learn (or be limited by) the flags of the particular tool I am using.

This is currently possible, though files need to be comma-separated due to how command line parsing is implemented:

`comby 'func' 'derp' $(find . -name "*.go" | tr '\n' ',')`

I may tinker with modifying it so that the `| tr '\n' ','` can be omitted.

**Comment by vogler** [comment]
Maybe also mention in the quick reference that .ext sets the matcher, or some warning when running without.
Just tried `comby 'let name =' 'let name () ='` and wondered why results are missing. Adding `.ml` solved it.
Strangely I couldn't see why the missing ones were not matched by the generic matcher since all matches seem to be the same.

**Comment by vogler** [comment]
> generic parser would fail to parse the OCaml comment
> parentheses are unbalanced. This could lead to, for example, missing matches for `let ...` in the same file downstream in the code.

Ah, ok, that's probably it. Thanks!

**Comment by rvantonder** [comment]
I've addressed these in the `0.8.0` cut. The new behavior is as follows:

- Warn loudly when the `generic` matcher is used or inferred, and give an informative message. The warning is suppressed when `-matcher .generic` is added to suppress the warning.
- Add a `-list` option that shows the matcher options for each supported language
- Infer languages from patterns like `go` (and not just `.go`, without the dot separator).

> Fastmod lets you put all the files you want at the end like grep.

I've removed the comma-separated parsing and comby now uses space-separated files/paths. So the `$(find ...)` expression works.

> removing all -directory, -extensions, -r, etc., flags to avoid confusion

These flags are honored when the input is a zip file (using the `-zip` option) and can filter directories/files where conventional linux utilities like `find` don't operate on zips/or are not available. 

> Maybe also mention in the quick reference that .ext sets the matcher

Updated in the site
